stubs/tensorflow/tensorflow/__init__.pyi

```diff
@@ -0,0 +1,195 @@
+_ShapeLike: TypeAlias = TensorShape | Iterable[_ScalarTensorCompatible | None] | int | Tensor
+_DTypeLike: TypeAlias = DType | str | numpy.dtype[Any]
+
+class Tensor:
+    def __init__(self, op: Operation, value_index: int, dtype: DType) -> None: ...
+    def consumers(self) -> list[Incomplete]: ...
+    @property
+    def shape(self) -> TensorShape: ...
+    def get_shape(self) -> TensorShape: ...
+    @property
+    def dtype(self) -> DType: ...
+    @property
+    def graph(self) -> Graph: ...
+    @property
+    def name(self) -> str: ...
+    @property
+    def op(self) -> Operation: ...
+    def numpy(self) -> numpy.ndarray[Any, Any]: ...
+    def __int__(self) -> int: ...
+    def __abs__(self, name: str | None = None) -> Tensor: ...
+    def __add__(self, other: _TensorCompatible) -> Tensor: ...
+    def __radd__(self, other: _TensorCompatible) -> Tensor: ...
+    def __sub__(self, other: _TensorCompatible) -> Tensor: ...
+    def __rsub__(self, other: _TensorCompatible) -> Tensor: ...
+    def __mul__(self, other: _TensorCompatible) -> Tensor: ...
+    def __rmul__(self, other: _TensorCompatible) -> Tensor: ...
+    def __pow__(self, other: _TensorCompatible) -> Tensor: ...
+    def __matmul__(self, other: _TensorCompatible) -> Tensor: ...
+    def __rmatmul__(self, other: _TensorCompatible) -> Tensor: ...
+    def __floordiv__(self, other: _TensorCompatible) -> Tensor: ...
+    def __rfloordiv__(self, other: _TensorCompatible) -> Tensor: ...
+    def __truediv__(self, other: _TensorCompatible) -> Tensor: ...
+    def __rtruediv__(self, other: _TensorCompatible) -> Tensor: ...
+    def __neg__(self, name: str | None = None) -> Tensor: ...
+    def __and__(self, other: _TensorCompatible) -> Tensor: ...
+    def __rand__(self, other: _TensorCompatible) -> Tensor: ...
+    def __or__(self, other: _TensorCompatible) -> Tensor: ...
+    def __ror__(self, other: _TensorCompatible) -> Tensor: ...
+    def __eq__(self, other: _TensorCompatible) -> Tensor: ...  # type: ignore[override]
+    def __ne__(self, other: _TensorCompatible) -> Tensor: ...  # type: ignore[override]
+    def __ge__(self, other: _TensorCompatible, name: str | None = None) -> Tensor: ...
+    def __gt__(self, other: _TensorCompatible, name: str | None = None) -> Tensor: ...
+    def __le__(self, other: _TensorCompatible, name: str | None = None) -> Tensor: ...
+    def __lt__(self, other: _TensorCompatible, name: str | None = None) -> Tensor: ...
+    def __bool__(self) -> NoReturn: ...
+    def __getitem__(self, slice_spec: _Slice | tuple[_Slice, ...]) -> Tensor: ...
+    def __len__(self) -> int: ...
+    # This only works for rank 0 tensors.
+    def __index__(self) -> int: ...
+    def __getattr__(self, name: str) -> Incomplete: ...
+
+class VariableSynchronization(Enum):
+    AUTO = 0
+    NONE = 1
+    ON_WRITE = 2
+    ON_READ = 3
+
+class VariableAggregation(Enum):
+    AUTO = 0
+    NONE = 1
+    ON_WRITE = 2
+    ON_READ = 3
```

Comment:
Generally speaking, we do our best to reproduce the structure of the runtime source code in typeshed's stubs. I don't see that much value in doing so with `tensorflow`, since I doubt you're meant to be importing these enums with `from tensorflow.python.ops.variables import VariableSynchronization`. But maybe we could add a comment, where possible, to indicate where these classes are actually defined at runtime?

Reply:
I'm hesitant as it's true for almost every single tensorflow class/function. Variable/Tensor are not defined in same file. Math functions are mostly split across files too. Similarly dtypes are not actually defined in that file. My guess is 95%+ of functions/classes the real definition is internal file and then re-exported elsewhere.

I can add a top level docstring to the file describing for where to find it? I usually find runtime class by going to [documentation](https://www.tensorflow.org/api_docs/python/tf/dtypes/DType) and clicking view source. 

A related thing is I'm not currently mimicking the full class inheritance. Mainly as most parent classes are undocumented internal classes. Metaclasses I only added for stubtest.

Reply:
Yeah that's all very fair. I guess this is just something I'll have to mentally adjust to with our TensorFlow stubs — "nothing will be where it seems" :)

> I can add a top level docstring to the file describing for where to find it? I usually find runtime class by going to [documentation](https://www.tensorflow.org/api_docs/python/tf/dtypes/DType) and clicking view source.

Good to know that it's easy to find the source from the docs! In that case, we can probably do without the docstrings in the stub.